walkthroughSteps/code-along-mentor-notes.md

```diff
@@ -3,56 +3,225 @@
+    ```js
+    const { Pool } = require('pg');
+    const url = require('url');
+    require('env2')('./config.env');
+
+    if (!process.env.DB_URL) throw new Error('Environment variable DB_URL must be set');
+    ```
+  
+    - `{ Pool }` is syntactic sugar (shorten/simplify syntax with abstraction) ([destructuring assignment](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)) that is equivalent to:
+        ```js
+        const pg = require('pg');
+        const Pool = pg.Pool;
+        ```
+    
+    - [`Connection pooling`](https://en.wikipedia.org/wiki/Connection_pool) is a cache of multiple database connections that are kept open for a timeout period (`idleTimeoutMillis`) and reused when future requests are required, minimising the resource impact of opening/closing connections constantly for write/read heavy apps. Reusing connections minimises latency too. Debug/demo logging `Pool` might be helpful.
+    - `url` is a Node module - `url.parse()` will be used later
+    - You'll notice that this file requires a `config.env`. We'll set this up later.
+    - This is a *gitignored* file with environment variables which are accessed with `process.env.NAME_HERE` and can be set in `config.env` *OR* production environments with `Heroku`
+    - The if statement will deliberately crash the script if the _connection information_ variable is missing
+
+4. Parse the URL and authentication info with this code:
+    ```js
+    const params = url.parse(process.env.DB_URL);
+    const [username, password] = params.auth.split(':');
+    ```
+    
+    - `url.parse(<url string here>)` will split a URL/HREF string into an object of values like `protocol`, `auth`, `hostname`, `port`: [URL split documentation](https://nodejs.org/api/url.html#url_url_strings_and_url_objects)
+    - `[username, password]` is a ES6 destructuring assignment that is syntactic sugar for:
+        ```js
+        const username = params.auth.split(':')[0];
+        const password = params.auth.split(':')[1];
+        ```
+    
+    Where username is index 0 of `params.auth.split(':')` and password is index 1, and so on.
+
+5. Create a [`pg options`](https://node-postgres.com/features/connecting#programmatic) object:
+    ```js
+    const options = {
+      host: params.hostname,
+      port: params.port,
+      database: params.pathname.split('/')[1],
+      max: process.env.DB_MAX_CONNECTIONS || 2,
+      user: username,
+      password,
+      ssl: params.hostname !== 'localhost',
+    }
+    ```
+    
+    - Use an appropriate number for `max`. More connections mean more memory is used, and too many can crash the database server. Always return connections to the pool (don't block/freeze query callbacks), or the pool will deplete. More connections mean more queries can be run at once and more redundancy incase connections are blocked/frozen.
+    - `ssl` will enable SSL (set to true) if you're not testing on a local machine.
```

Comment:
Add that SSL (or secure sockets layer) when set to true ensures you are connecting to the database from a secure server (https)